versioned_hdf5/slicetools.pyx

```diff
@@ -191,3 +194,253 @@ cpdef build_data_dict(dcpl, raw_data_name: str):
             free(filename_buf)
 
     return data_dict
+
+
+@cython.cfunc
+@cython.returns(cython.Py_ssize_t)
+def _ceiling(a: cython.Py_ssize_t, b: cython.Py_ssize_t):
+    """
+    Returns ceil(a/b)
+    """
+    return -(-a//b)
+
+@cython.cfunc
+@cython.returns(cython.Py_ssize_t)
+def _min(a: cython.Py_ssize_t, b: cython.Py_ssize_t):
+    return min(a, b)
+
+@cython.cfunc
+@cython.returns(cython.Py_ssize_t)
+def _smallest(x: cython.Py_ssize_t, a: cython.Py_ssize_t, m: cython.Py_ssize_t):
+    """
+    Gives the smallest integer >= x that equals a (mod m)
+
+    Assumes x >= 0, m >= 1, and 0 <= a < m.
+    """
+    n: cython.Py_ssize_t = _ceiling(x - a, m)
+    return a + n*m
+
+@cython.ccall
+def subindex_chunk_slice(c_start: cython.Py_ssize_t, c_stop: cython.Py_ssize_t,
+                         i_start: cython.Py_ssize_t, i_stop: cython.Py_ssize_t, i_step: cython.Py_ssize_t):
+    common: cython.Py_ssize_t = i_start % i_step
+
+    lcm: cython.Py_ssize_t = i_step
+    start: cython.Py_ssize_t = max(c_start, i_start)
+
+    # Get the smallest lcm multiple of common that is >= start
+    start = _smallest(start, common, lcm)
+    # Finally, we need to shift start so that it is relative to index
+    start = (start - i_start)//i_step
+
+    stop: cython.Py_ssize_t = _ceiling((_min(c_stop, i_stop) - i_start), i_step)
+    stop = 0 if stop < 0 else stop
```

Comment:
Integer overflow: all these variables are unsigned integers so if there is a chance of `i_start >= min(c_stop, i_stop)` this will break.
```suggestion
    stop: cython.Py_ssize_t = min(c_stop, i_stop)
    stop = _ceiling(stop - i_start, i_step) if stop > i_start else 0
```